### PR TITLE
feat(version): Set and surface CLI version.

### PR DESCRIPTION
**Pull request by jtk54**

### CODE REVIEW

**Comment by lwander** [comment]
Do we hard-code the release version, or is it set automatically during a release?

**Comment by jtk54** [comment]
We'd set it during a build/release cycle: https://github.com/spinnaker/spinnaker/pull/3182.

File: version/version.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	"fmt"
+)
+
+// The git commit that was compiled. This is filled in by the compiler.
+var GitCommit string
+
+// Main version number being run right now.
+var Version = "0.4.0"
```
Comment (ewiseblatt) [inline version/version.go]: Why is version information hardcoded in the source?
It is usually imposed from the outside. If the compiler is filling in the GitCommit,
why cant you also fill in the version (and release phase below)?

File: version/version.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+// A pre-release marker for the version. If this is an empty string,
+// then the release is a final release. Otherwise this is a pre-release
+// version e.g. "dev", "alpha", etc.
+var Prerelease = "dev"
```
Comment (ewiseblatt) [inline version/version.go]: Why not give this a name like ReleasePhase? You can either special case "final" to encode as "" or special case "" to mean final. If you go the encoding route, you can have a function that returns the encoding decorator for the release phase (which would either be "" or "-<phase>" depending on phase being "final" or not)

File: command/meta.go [hunk 3]
```diff
@@ -160,7 +161,7 @@ func (m *ApiMeta) Process(args []string) ([]string, error) {
 	cfg := &gate.Configuration{
 		BasePath:      m.gateEndpoint,
 		DefaultHeader: make(map[string]string),
-		UserAgent:     "Spin CLI version", // TODO(jacobkiefer): Add a reasonable UserAgent.
+		UserAgent:     fmt.Sprintf("%s: %s", version.Header, version.VersionString()),
```
Comment (ewiseblatt) [inline command/meta.go]: sorry, not familiar with programming model here and github UI isnt helpful.
Is the UserAgent the HTTP header? It should be in the format "<program>/<version> ..."
I'm not sure what version.Header is here. The name "header" is odd.

File: version/version.go (hunk 1 above)
Comment (jtk54) [reply version/version.go]: We'd set it during a build/release cycle: spinnaker/spinnaker#3182 (look for `ldflags`). It's set here also so we have a reasonable user agent and version for built versions that haven't been released yet.

File: version/version.go (hunk 2 above)
Comment (jtk54) [reply version/version.go]: `ReleasePhase` makes sense.

File: command/meta.go (hunk 3 above)
Comment (jtk54) [reply command/meta.go]: Not tied to the name. Is the "/..." common? Something like `/Spin CLI: <version>`?

**Comment by jtk54** [comment]
Unset the hardcoded version, added a script to support building from source if folks want to do that.

File: build.sh [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+if [ ! -z "$go_os"]; then
+    export $GO_OS=$go_os
+fi
+
+LD_FLAGS="-X github.com/spinnaker/spin/version.Version=${VERSION} \
+-X github.com/spinnaker/spin/version.ReleasePhase=${RELEASE_PHASE}"
+go clean
+go get -d -v
+env CGO_ENABLED=0 go build -ldflags "${LD_FLAGS}" .
```
Comment (brandonnelson3) [inline build.sh]: I think "env" is unneccessary.

File: version/version.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+// A pre-release marker for the version. If this is an empty string,
+// then the release is a final release. Otherwise this is a pre-release
+// version e.g. "dev", "alpha", etc.
+var ReleasePhase = "dev"
+
+// User Agent name set in requests.
+const UserAgent = "Spin-CLI"
+
+// VersionString prints the version of the Spin CLI.
+func VersionString() string {
```
Comment (brandonnelson3) [inline version/version.go]: Version is redundant with the package, consider just String(), so external usage reads version.String().

**Pull request merged**